[mlir][ArmSME] Support lowering masked vector.outerproduct ops to SME

MacDue opened the pull request:
This patch adds support for lowering masked outer products to SME. This is done in two stages. First, vector.outerproducts (both masked and non-masked) are rewritten to arm_sme.outerproducts. The arm_sme.outerproduct op is close to vector.outerproduct, but supports masking on the operands rather than the result. It also limits the cases it handles to things that could be (directly) lowered to SME.

This currently requires that the source of the mask is a vector.create_mask op. E.g.:

```mlir
%mask = vector.create_mask %dimA, %dimB : vector<[4]x[4]xi1>
%result = vector.mask %mask {
             vector.outerproduct %vecA, %vecB
              : vector<[4]xf32>, vector<[4]xf32>
          } : vector<[4]x[4]xi1> -> vector<[4]x[4]xf32>
```
Is rewritten to:
```
%maskA = vector.create_mask %dimA : vector<[4]xi1>
%maskB = vector.create_mask %dimB : vector<[4]xi1>
%result = arm_sme.outerproduct %vecA, %vecB masks(%maskA, %maskB)
              : vector<[4]xf32>, vector<[4]xf32>
```
(The same rewrite works for non-masked vector.outerproducts too)

The arm_sme.outerproduct can then be directly lowered to SME intrinsics.

MacDue committed:
[mlir][ArmSME] Support lowering masked vector.outerproduct ops to SME

This patch adds support for lowering masked outer products to SME. This
is done in two stages. First, vector.outerproducts (both masked and
non-masked) are rewritten to arm_sme.outerproducts. The
arm_sme.outerproduct op is close to vector.outerproduct, but supports
masking on the operands rather than the result. It also limits the
cases it handles to things that could be lowered to SME, but does not
enforce that everything matches SME tiles at this level.

This currently requires that the source of the mask is a
vector.create_mask op. E.g.:

```mlir
%mask = vector.create_mask %dimA, %dimB : vector<[4]x[4]xi1>
%result = vector.mask %mask {
             vector.outerproduct %vecA, %vecB
              : vector<[4]xf32>, vector<[4]xf32>
          } : vector<[4]x[4]xi1> -> vector<[4]x[4]xf32>
```
Is rewritten to:
```
%maskA = vector.create_mask %dimA : vector<[4]xi1>
%maskB = vector.create_mask %dimB : vector<[4]xi1>
%result = arm_sme.outerproduct %vecA, %vecB masks(%maskA, %maskB)
              : vector<[4]xf32>, vector<[4]xf32>, vector<[4]x[4]xf32>
```
(The same rewrite works for non-masked vector.outerproducts too)

The arm_sme.outerproduct can then be directly lowered to SME intrinsics.

MacDue commented:
cc @cfRod

MacDue committed:
Make comment in VectorToArmSME.cpp match others in style

MacDue commented on mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:
```diff
@@ -507,4 +524,95 @@ def MoveTileSliceToVectorOp : ArmSME_Op<"move_tile_slice_to_vector", [Pure,
 }
 
+class HasMatchingMaskTypeConstraint<string operand, string maskGetter> :
+  TypesMatchWith<
+    "shape of `" # operand #  "Mask` matches `" # operand # "`",
+    operand, operand # "Mask",
+    "::llvm::cast<mlir::VectorType>($_self).cloneWith({}, IntegerType::get($_ctxt, 1))",
+    "!" # maskGetter # "() || std::equal_to<>()">;
```
TODO: Use `OptionalTypesMatchWith` here

MacDue commented on mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:
```diff
@@ -507,4 +524,95 @@ def MoveTileSliceToVectorOp : ArmSME_Op<"move_tile_slice_to_vector", [Pure,
+          "getLhsType().getDimSize(0), getRhsType().getDimSize(0)},"
+          "getRhsType().getElementType(),"
+          "{getLhsType().getScalableDims()[0], getRhsType().getScalableDims()[0]})">>,
+    TypesMatchWith<"`result` and `acc` have the same type",
+      "result", "acc",
+      "::llvm::cast<mlir::Type>($_self)",
+      "!getAcc() || std::equal_to<>()">
```
...and here

MacDue committed:
Switch to OptionalTypesMatchWith

banach-space reviewed (commented):
This is looking great! Mostly nits from me, but I need another round :)

dcaballe reviewed (approved):
Cool!

MacDue commented on mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:
```diff
@@ -79,6 +79,23 @@ def ArmSME_TileSliceLayoutAttr : EnumAttr<ArmSME_Dialect, TileSliceLayout,
+}
+
+/// An attribute that specifies how to combine a newly produced value with the
+/// accumulator. This is similar to vector::CombiningKindAttr, but limited to
```
I did try this, but including `VectorAttributes.td` within this file results in the generator creating duplicate definitions of attributes from the vector dialect (which fails to build).

MacDue commented on mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:
```diff
@@ -507,4 +524,93 @@ def MoveTileSliceToVectorOp : ArmSME_Op<"move_tile_slice_to_vector", [Pure,
+    This op is based on `vector.outerproduct` with the extra conditions that:
+
+    * AXPY operations are not supported
+    * The only combining functions are "add" and "sub"
```
We've not yet implemented `sub` (or the intrinsics required for it yet).

MacDue commented on mlir/test/Integration/Dialect/Vector/CPU/ArmSME/test-outerproduct-f64.mlir:
```diff
@@ -11,6 +11,9 @@
 
 // RUN: %{compile} | %{run} | FileCheck %s
 
+// REDEFINE: %{entry_point} = test_masked_outerproduct_with_accumulator_2x2xf64
```
These are both tests with accumulators, there's no non-accumulator tests in this file. Not sure why that is, but I could add some.

MacDue replied:
I've now extended the f64 tests to match the f32 ones (no acc, with acc, masked no acc, masked with acc).

MacDue commented on mlir/lib/Conversion/VectorToArmSME/VectorToArmSME.cpp:
```diff
@@ -427,12 +427,88 @@ struct TransposeOpToArmSMELowering
+                                PatternRewriter &rewriter) const override {
+    // AXPY operation not suited for SME.
+    if (!isa<VectorType>(outerProductOp.getOperandTypeRHS()))
+      return outerProductOp.emitError("AXPY operations not supported");
```
If does support if you want it... You'd just have to mask out all but the first element of the LHS. We're choosing not to support it though.

MacDue commented on mlir/test/Dialect/ArmSME/vector-ops-to-llvm.mlir:
```diff
@@ -463,17 +463,82 @@ func.func @vector_outerproduct_no_accumulator(%lhs : vector<[2]xf64>, %rhs : vec
+// CHECK-LABEL: @vector_outerproduct_masked_f32
+// CHECK-SAME: (%[[LHS:.*]]: vector<[4]xf32>, %[[RHS:.*]]: vector<[4]xf32>, %[[ACC:.*]]: vector<[4]x[4]xf32>,
+func.func @vector_outerproduct_masked_f32(%lhs : vector<[4]xf32>, %rhs : vector<[4]xf32>, %acc : vector<[4]x[4]xf32>, %dim0: index, %dim1: index) {
+  // CHECK: %[[LHS_MASK:.*]] = arith.cmpi slt, {{.*}} : vector<[4]xi32>
```
I'll add checks for the first test, it's somewhat long, so I'll keep the shorthand for the other tests.

MacDue committed:
Fixups

This also adds some no accumulator tests to test-outerproduct-f64.mlir
(and removes some now done TODOs, e.g. vector.splat support).

banach-space commented on mlir/test/Dialect/ArmSME/invalid.mlir:
```diff
@@ -97,3 +97,32 @@ func.func @arm_sme_move_tile_slice_to_vector__bad_result_type(%tile : vector<[4]
+
+// -----
+
+// expected-note@+1 {{prior use here}}
+func.func @arm_sme_outproduct__bad_mask_type(%vecA: vector<3xf32>, %vecB: vector<[2]xf32>, %maskA: vector<5xi1>, %maskB: vector<[2]xi1>)-> vector<3x[2]xf32>
+{
+  // expected-error@+1 {{use of value '%maskA' expects different type than prior uses}}
+  %0 = arm_sme.outerproduct %vecA, %vecB masks(%maskA, %maskB) : vector<3xf32>, vector<[2]xf32>, vector<3x[2]xf32>
+  return %0 : vector<3x[2]xf32>
+}
```
Hm, isn't this a generic MLIR error rather than something specific to `arm_sme.outerpoduct`?

MacDue replied:
It is an attempt to test the constraint used to infer the types of these masks, I couldn't figure out how to get it to use the custom error.

banach-space commented on mlir/test/Dialect/ArmSME/invalid.mlir:
```diff
@@ -97,3 +97,32 @@ func.func @arm_sme_move_tile_slice_to_vector__bad_result_type(%tile : vector<[4]
+
+// -----
+
+// expected-note@+1 {{prior use here}}
+func.func @arm_sme_outproduct__bad_acc_type(%vecA: vector<7xi32>, %vecB: vector<6xi32>, %acc: vector<6x6xi32>) -> vector<7x6xi32>
+{
+  // expected-error@+1 {{use of value '%acc' expects different type than prior uses}}
+  %0 = arm_sme.outerproduct %vecA, %vecB acc(%acc) : vector<7xi32>, vector<6xi32>, vector<7x6xi32>
+  return %0 : vector<7x6xi32>
+}
```
Also here it feels like we are testing generic MLIR functionality rather than something specific to `arm_sme.outerproduct`?

MacDue replied:
Again this shows the constraint used to infer the type of `acc` works, but the custom error message is not used (for some reason).

banach-space commented on mlir/test/Dialect/ArmSME/invalid.mlir:
```diff
@@ -97,3 +97,32 @@ func.func @arm_sme_move_tile_slice_to_vector__bad_result_type(%tile : vector<[4]
+
+// -----
+
+func.func @arm_sme_outproduct__bad_result_type(%vecA: vector<[8]xi16>, %vecB: vector<[8]xi16>) -> vector<[2]x1xi16>
+{
+  // expected-error@+1 {{op failed to verify that result type is derived from `lhs` and `rhs`}}
+  %0 = arm_sme.outerproduct %vecA, %vecB : vector<[8]xi16>, vector<[8]xi16>, vector<[2]x1xi16>
+  return %0 : vector<[2]x1xi16>
+}
```
How about:
```
  %0 = arm_sme.outerproduct %vecA, %vecB : vector<[8]xi16>, vector<[8]xi16>, vector<8x8xi16>
  %1 = arm_sme.outerproduct %vecA, %vecB : vector<[8]xi16>, vector<[8]xi16>, vector<[8]x8xi16>
  %2 = arm_sme.outerproduct %vecA, %vecB : vector<[8]xi16>, vector<[8]xi16>, vector<[4]x[4]xi16>
  %3 = arm_sme.outerproduct %vecA, %vecB : vector<[8]xi16>, vector<[8]xi16>, vector<[4]x[4]xi32>
```

MacDue replied:
N/A now (as the result is now inferred)

banach-space commented on mlir/test/Dialect/ArmSME/roundtrip.mlir:
```diff
@@ -1151,3 +1151,56 @@ func.func @arm_sme_move_tile_slice_to_vector_ver_f64(%tile : vector<[2]x[2]xf64>
+// -----
+
+func.func @arm_sme_outproduct_with_masking(
+  %vecA: vector<3xf32>, %vecB: vector<[2]xf32>, %maskA: vector<3xi1>, %maskB: vector<[2]xi1>
+) -> vector<3x[2]xf32> {
```
[nit] I think that you can safely keep these in one line (I personally prefer consistent formatting over wrapping lines at e.g. 80 chars)

banach-space commented on mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:
```diff
@@ -515,4 +536,93 @@ def MoveTileSliceToVectorOp : ArmSME_Op<"move_tile_slice_to_vector", [Pure,
+  let description = [{
+    This op is based on `vector.outerproduct` with the extra conditions that:
+
+    * AXPY operations are not supported
```
Is this left as a TODO? Worth mentioning.

MacDue replied:
I'm not sure if it's a TODO? I don't know if we want to support it :)

MacDue replied:
I've left this as a comment in the lowering now (and rewrote the description due to other changes).

banach-space commented on mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:
```diff
@@ -515,4 +536,93 @@ def MoveTileSliceToVectorOp : ArmSME_Op<"move_tile_slice_to_vector", [Pure,
+    * The only combining functions are "add" and "sub"
+    * Masking is performed on the inputs (rather than the output)
+
+    This is meant as an intermediate op for lowering `vector.outerproduct` to
+    SME. Types are not restricted to SVE/SME vectors at this level.
```
What happens if this condition is violated and somebody tries to lower to LLVM?

MacDue replied:
Nothing good! I've limited things to SME tiles now.

banach-space commented on mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:
````diff
@@ -515,4 +536,93 @@ def MoveTileSliceToVectorOp : ArmSME_Op<"move_tile_slice_to_vector", [Pure,
+                : vector<[4]xf32>, vector<[4]xf32>, vector<[4]x[4]xf32>
+    ```
+
+    Example 2: Unmasked outerproduct (with accumulator)
+    ```mlir
+    %result = arm_sme.outerproduct $lhs, $rhs acc($accumulator)
+                : vector<[4]xf32>, vector<[4]xf32>, vector<[4]x[4]xf32>
+    ```
````
What are the semantics if the accumulator is not present? Will the lowering take care of initializing to 0? Worth mentioning.

MacDue replied:
It's zeroing :)

banach-space commented on mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:
```diff
@@ -515,4 +536,93 @@ def MoveTileSliceToVectorOp : ArmSME_Op<"move_tile_slice_to_vector", [Pure,
+  }];
+
+  let arguments = (ins
+    VectorOfRank<[1]>:$lhs, VectorOfRank<[1]>:$rhs,
```
Why wouldn't we restrict this to SVE vector?

banach-space commented on mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:
```diff
@@ -515,4 +536,93 @@ def MoveTileSliceToVectorOp : ArmSME_Op<"move_tile_slice_to_vector", [Pure,
+    VectorOfRank<[1]>:$lhs, VectorOfRank<[1]>:$rhs,
+    Optional<VectorOfRankAndType<[1],[I1]>>:$lhsMask,
+    Optional<VectorOfRankAndType<[1],[I1]>>:$rhsMask,
+    Optional<VectorOfRank<[2]>>: $acc,
```
Why not restrict this to an SME tile type?

banach-space commented on mlir/lib/Conversion/VectorToArmSME/VectorToArmSME.cpp:
```diff
@@ -427,12 +427,88 @@ struct TransposeOpToArmSMELowering
+
+  LogicalResult matchAndRewrite(vector::OuterProductOp outerProductOp,
+                                PatternRewriter &rewriter) const override {
+    // AXPY operation not suited for SME.
```
Not suited or not supported yet? It could be supported with masking, right?

MacDue replied:
I just took this from the previous outer product lowering. You can support it with making :)

banach-space commented on mlir/lib/Conversion/VectorToArmSME/VectorToArmSME.cpp:
```diff
@@ -427,12 +427,88 @@ struct TransposeOpToArmSMELowering
+
+    auto kind = outerProductOp.getKind();
+    if (kind != vector::CombiningKind::ADD)
+      return outerProductOp.emitError("unsupported kind");
```
Perhaps something more descriptive:
```
"unsupported kind (lowering to SME only supports ADD at the moment)"
```

banach-space commented on mlir/lib/Conversion/VectorToArmSME/VectorToArmSME.cpp:
```diff
@@ -427,12 +427,88 @@ struct TransposeOpToArmSMELowering
 
+/// Conversion pattern for vector.outerproduct.
+///
+/// If the vector.outerproduct is masked (and the mask from a
+/// vector.create_mask), then the mask is decomposed into two 1-D masks for the
+/// operands.
```
And what if it isn't masked?

MacDue replied:
s/vector.outerproduct/arm_sme.outerproduct/g :P

banach-space commented on mlir/lib/Conversion/VectorToArmSME/VectorToArmSME.cpp:
```diff
@@ -427,12 +427,88 @@ struct TransposeOpToArmSMELowering
+        return failure();
+
+      auto maskType = createMaskOp.getVectorType();
+      if (maskType.getRank() != 2)
```
This should never happen, right? As in, that would be an invalid `vector.outerproduct`? Just checking my understanding.

MacDue replied:
Yep (I've removed this check now, as it's clearly redundant now that only tile-sized operations are converted to `arm_sme.outerproduct`).

banach-space commented on mlir/lib/Conversion/VectorToArmSME/VectorToArmSME.cpp:
```diff
@@ -427,12 +427,88 @@ struct TransposeOpToArmSMELowering
+    Value lhsMask = {};
+    Value rhsMask = {};
+    Operation *rootOp = outerProductOp;
+    if (outerProductOp.isMasked()) {
+      auto maskingOp = outerProductOp.getMaskingOp();
+      rewriter.setInsertionPoint(maskingOp);
+      rootOp = maskingOp;
+
+      // Attempt to extract masks from vector.create_mask.
+      // TODO: Add support for other mask sources.
+      auto mask = maskingOp.getMask();
+      auto createMaskOp = mask.getDefiningOp<vector::CreateMaskOp>();
+      if (!createMaskOp)
+        return failure();
+
+      auto maskType = createMaskOp.getVectorType();
+      if (maskType.getRank() != 2)
+        return failure();
+
+      auto loc = outerProductOp.getLoc();
+
+      Value lhsMaskDim = createMaskOp.getOperand(0);
+      Value rhsMaskDim = createMaskOp.getOperand(1);
+
+      VectorType operandMaskType = VectorType::Builder(maskType).dropDim(0);
+      lhsMask = rewriter.create<vector::CreateMaskOp>(loc, operandMaskType,
+                                                      lhsMaskDim);
+      rhsMask = rewriter.create<vector::CreateMaskOp>(loc, operandMaskType,
+                                                      rhsMaskDim);
+    }
+
+    rewriter.replaceOpWithNewOp<arm_sme::OuterProductOp>(
+        rootOp, outerProductOp.getResultVectorType(), outerProductOp.getLhs(),
+        outerProductOp.getRhs(), lhsMask, rhsMask, outerProductOp.getAcc());
+
+    return success();
```
Try to avoid big blocks of indentation: https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code

```suggestion
    if (!outerProductOp.isMasked()) {
        rewriter.replaceOpWithNewOp<arm_sme::OuterProductOp>(
        rootOp, outerProductOp.getResultVectorType(), outerProductOp.getLhs(),
        outerProductOp.getRhs(), lhsMask, rhsMask, outerProductOp.getAcc());

        return success();
    }
     
    auto maskingOp = outerProductOp.getMaskingOp();
    rewriter.setInsertionPoint(maskingOp);
    rootOp = maskingOp;

    // Attempt to extract masks from vector.create_mask.
    // TODO: Add support for other mask sources.
    auto mask = maskingOp.getMask();
    auto createMaskOp = mask.getDefiningOp<vector::CreateMaskOp>();
    if (!createMaskOp)
      return failure();

    auto maskType = createMaskOp.getVectorType();
    if (maskType.getRank() != 2
      return failure();

    auto loc = outerProductOp.getLoc();

    Value lhsMaskDim = createMaskOp.getOperand(0);
    Value rhsMaskDim = createMaskOp.getOperand(1);

    VectorType operandMaskType = VectorType::Builder(maskType).dropDim(0);
    lhsMask = rewriter.create<vector::CreateMaskOp>(loc, operandMaskType,
                                                    lhsMaskDim);
    rhsMask = rewriter.create<vector::CreateMaskOp>(loc, operandMaskType,
                                                    rhsMaskDim);
  }
```

MacDue replied:
I've refactored this into a `decomposeResultMask()` helper function, which I think is nicer :)

banach-space reviewed (commented):
This is obviously great :) I've made a few suggestions and raised a few questions, but nothing major.

I do think that this could be split into two patches though:
* introduce `arm_sme.outerproduct` (and rewire the lowering for `vector.outerproduct`,
* add support for masking.

I'm not going to push for that, as Diego has already accepted and in general this very well crafted - the noise is not too bad. Just a note for our futures selves :)

MacDue committed:
Rework a little

The arm_sme.outerproduct op has now been updated to match SME tile sizes,
which cleans things up a little. A few other fixups included.

MacDue committed:
Fixups

MacDue committed:
Tidy up rewrite

banach-space commented on mlir/test/Integration/Dialect/Vector/CPU/ArmSME/test-outerproduct-f32.mlir:
```diff
@@ -14,6 +14,12 @@
 // REDEFINE: %{entry_point} = test_outerproduct_with_accumulator_4x4xf32
 // RUN: %{compile} | %{run} | FileCheck %s --check-prefix=WITH-ACC
```
[nit] It would be good to only compile once, see e.g. https://github.com/llvm/llvm-project/blob/640274ff1fe17852615564a80d9f8c69b67f1d59/mlir/test/Integration/Dialect/Vector/CPU/ArmSVE/test-contraction.mlir#L1-L20

banach-space reviewed (approved):
LGTM, thank you for addressing my comments :)

c-rhodes commented on mlir/test/Dialect/ArmSME/vector-ops-to-sme.mlir:
```diff
@@ -578,3 +578,99 @@ func.func @transpose_f64(%arg0: vector<[2]x[2]xf64>) {
+
+// -----
+
+// CHECK-LABEL: @vector_outerproduct_masked_f64
```
nit: everywhere else the tests are ordered smallest to largest (f16, bf16, f32, f64). It could just as well be largest to smallest, that's not particularly important, but I think consistency is. I'd appreciate it if we could be consistent, it makes reviewing easier.

MacDue replied:
Done (other than keeping the main example first)

c-rhodes reviewed (approved):
great work Ben, I've left a few comments all minor, otherwise LGTM, cheers

MacDue committed:
Only compile tests once

MacDue committed:
Fixups

MacDue commented on mlir/lib/Dialect/ArmSME/Transforms/LegalizeForLLVMExport.cpp:
```diff
@@ -474,13 +474,13 @@ struct MoveTileSliceToVectorArmSMELowering
-struct VectorOuterProductToArmSMELowering
-    : public ConvertOpToLLVMPattern<vector::OuterProductOp> {
-  using ConvertOpToLLVMPattern<vector::OuterProductOp>::ConvertOpToLLVMPattern;
+struct OuterProductToArmSMELowering
```
I've changed this to `OuterProductOpConversion` (which matches `ZeroOpConversion` now) :+1:

MacDue commented on mlir/test/Dialect/ArmSME/vector-ops-to-sme.mlir:
```diff
@@ -578,3 +578,99 @@ func.func @transpose_f64(%arg0: vector<[2]x[2]xf64>) {
+// CHECK-LABEL: @vector_outerproduct_masked_f64
+// CHECK-SAME: (%[[LHS:.*]]: vector<[2]xf64>, %[[RHS:.*]]: vector<[2]xf64>, %[[ACC:.*]]: vector<[2]x[2]xf64>, %[[DIM0:.*]]: index, %[[DIM1:.*]]: index
+func.func @vector_outerproduct_masked_f64(%lhs : vector<[2]xf64>, %rhs : vector<[2]xf64>, %acc : vector<[2]x[2]xf64>, %dim0: index, %dim1: index) {
+  %mask = vector.create_mask %dim0, %dim1 : vector<[2]x[2]xi1>
```
This style is used in many places, and I find nicer to read :)
